### ISSUE TITLE
No Shapely geometry can be created from null value

### ISSUE THREAD

**Issue by gnthibault** [shapely/shapely#820]
## Expected behavior and actual behavior.

Expected behaviour: no error
Actual behaviour:

```
TopologyException: assigned depths do not match at 882.50000000000011 1241.5
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-69-fe56313cbba3> in <module>
----> 1 line.parallel_offset(distance=2.5, side='left')

~/anaconda3/lib/python3.6/site-packages/shapely/geometry/linestring.py in parallel_offset(self, distance, side, resolution, join_style, mitre_limit)
    144         try:
    145             return geom_factory(self.impl['parallel_offset'](
--> 146                 self, distance, resolution, join_style, mitre_limit, side))
    147         except OSError:
    148             raise TopologicalError()

~/anaconda3/lib/python3.6/site-packages/shapely/geometry/base.py in geom_factory(g, parent)
     74     # Abstract geometry factory for use with topological methods below
     75     if not g:
---> 76         raise ValueError("No Shapely geometry can be created from null value")
     77     ob = BaseGeometry()
     78     geom_type = geometry_type_name(g)

ValueError: No Shapely geometry can be created from null value

```
## Steps to reproduce the problem.

```python
import shapely
line = shapely.geometry.linestring.LineString([(884.0, 1240.0),(883.0, 1239.0),(882.0, 1239.0),(881.0, 1240.0)])                                                                                   
line.parallel_offset(distance=2.5, side='left') 
```

The weird thing is that the problem does not show if distance=2.499999999999 or distance=2.5000000000001

## Operating system

Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.6 LTS
Release:	16.04
Codename:	xenial

## Shapely version and provenance

Shapely==1.6.4.post2 installed from PyPI using pip

**Comment by sgillies** [comment]
@gnthibault would you be willing to try to reproduce the problem using shapely==1.7a3 and shapely==1.7b1? I have a feeling this is an upstream bug that has been fixed.

**Comment by Jeremiah-England** [comment]
I can reproduce this with python 3.8.2, shapely==1.7.1 on a  WSL using Ubuntu 20.04.1, 

```python
from shapely.geometry import LineString
line = LineString([(884.0, 1240.0),(883.0, 1239.0),(882.0, 1239.0),(881.0, 1240.0)])
print(line.parallel_offset(distance=2.5, side='left').wkt)
```

```python-traceback
TopologyException: assigned depths do not match at 882.50000000000011 1241.5
Traceback (most recent call last):
  File "shapely_890.py", line 3, in <module>
    print(line.parallel_offset(distance=2.5, side='left').to_wkt())
  File "/home/visiostack/.local/lib/python3.8/site-packages/shapely/geometry/linestring.py", line 145, in parallel_offset
    return geom_factory(self.impl['parallel_offset'](
  File "/home/visiostack/.local/lib/python3.8/site-packages/shapely/geometry/base.py", line 78, in geom_factory
    raise ValueError("No Shapely geometry can be created from null value")
ValueError: No Shapely geometry can be created from null value
```

It does not happen for `2.5000000000001`, but it does for `2.50000000000001`.

**Comment by Jeremiah-England** [comment]
I ran into another example of this issue, only it has a different topology exception underneath:

```python
import shapely.wkt
shapely.wkt.loads('LINESTRING (12 20, 60 68, 111 114, 151 159, 210 218)').parallel_offset(3)
```

```python-traceback
TopologyException: found non-noded intersection between LINESTRING (25.1837 28.941, 25.1837 28.941) and LINESTRING (25.1837 28.941, 25.1837 28.941) at 25.183658239422783 28.941017552303521

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-21-df893580dce5> in <module>
----> 1 shapely.wkt.loads('LINESTRING (12 20, 60 68, 111 114, 151 159, 210 218)').parallel_offset(3)

~/.local/lib/python3.8/site-packages/shapely/geometry/linestring.py in parallel_offset(self, distance, side, resolution, join_style, mitre_limit)
    143                 'Cannot compute offset from zero-length line segment')
    144         try:
--> 145             return geom_factory(self.impl['parallel_offset'](
    146                 self, distance, resolution, join_style, mitre_limit, side))
    147         except OSError:

~/.local/lib/python3.8/site-packages/shapely/geometry/base.py in geom_factory(g, parent)
     76     # Abstract geometry factory for use with topological methods below
     77     if not g:
---> 78         raise ValueError("No Shapely geometry can be created from null value")
     79     ob = BaseGeometry()
     80     geom_type = geometry_type_name(g)

ValueError: No Shapely geometry can be created from null value
```

**Comment by jorisvandenbossche** [comment]
With latest main branch, we correctly raise the GEOS error instead of the confusing "No Shapely geometry can be created from null value" error:

```
>>> from shapely.geometry import LineString
>>> line = LineString([(884.0, 1240.0),(883.0, 1239.0),(882.0, 1239.0),(881.0, 1240.0)])
>>> line.parallel_offset(distance=2.5, side='left')
...
GEOSException: TopologyException: assigned depths do not match at 882.50000000000011 1241.5
```

So I think we label / close this as an upstream GEOS issue?

**Comment by jorisvandenbossche** [comment]
In addition, with GEOS 3.11, this now works without exception:

```
>>> from shapely.geometry import LineString
>>> line = LineString([(884.0, 1240.0),(883.0, 1239.0),(882.0, 1239.0),(881.0, 1240.0)])
>>> line.parallel_offset(distance=2.5, side='left')
<LINESTRING (885.768 1238.232, 884.768 1237.232, 884.586 1237.067, 884.389 1...>
```

GEOS 3.11 will be included in the shapely 2.0 wheels (and with conda you also already use this with shapely 1.8.5).

Closing as fixed upstream in GEOS.
